### PR TITLE
Implement one-shot functionality

### PR DESCRIPTION
**Pull request by jtroo**

### CODE REVIEW

**Comment by jtroo** [comment]
This is for #62

**Comment by jtroo** [comment]
I found an issue where double-pressing the one-shot key causes the key to remain released even while still held. I will push a fix shortly.

**Comment by jtroo** [comment]
> I found an issue where double-pressing the one-shot key causes the key to remain released even while still held. I will push a fix shortly.

Issue is fixed and a test added.

File: src/action.rs [hunk 1]
```diff
@@ -186,6 +186,23 @@ where
...
+    /// one shot activation or `timeout` ticks elapse. For example, a one-shot key can be
+    /// used to activate shift for exactly one keypress or switch to another layer for
+    /// exactly one keypress.
+    ///
+    /// Holding a one-shot key will be treated as a normal held keypress.
+    ///
+    /// If you use one-shot outside of its intended use cases (modifier key action or layer
+    /// action) then you will likely have undesired behaviour. E.g. one-shot with the space
+    /// key will hold space until either another key is pressed or the timeout occurs, which will
+    /// probably send many undesired space characters to your active application.
```
Comment (TeXitoi) [inline src/action.rs:199]: what about multiple one shot? layer one shot? That's important feature IMHO. Allowing something like:
* one shot alt
* one shot ctrl
* one shot action layer
* suppr (on the action layer)
would be very soon needed.

Comment (TeXitoi) [reply src/action.rs:199]: Yeah, multiple one shot.  Good for layer (so it works for any action, but might do strange things) .

**Comment by TeXitoi** [comment]
Maybe the release of the one shot should be at the first not one shot release, allowing to:  
 * one shot Ctrl
 * press alt
 * press suppr

Or maybe not? (it might cause one shot shift + rowling not doing the intented behavior).

**Comment by TeXitoi** [comment]
What's the behavior in QMK and ZMK? If that's it for the 2, I'm OK with it.

**Comment by jtroo** [comment]
From reading the documentation:

- QMK doesn't advertise combining multiple one shot mods (though it supports it according to https://github.com/qmk/qmk_firmware/issues/2796)
- ZMK allows combining multiple sticky keys
- ZMK allows configuring if a press or release ends its sticky keys (quick-release) setting via `quick-release`
  - It probably wouldn't be too hard to add this, so may as well

**Comment by TeXitoi** [comment]
What about:
```rust
    OneShot(&'static OneShot), // for the small action object, see previous PR

// ...

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct OneShot<T> {
    /// Action to activate until timeout expires or exactly one keypress is activated.
    action: Action<T>,
    /// Timeout after which one-shot will expire.
    timeout: u16,
    /// configuration of the one shot behavior
    config: OneShotConfig
}

#[non_exhaustive]
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum OneShotConfig {
    OnFirstPress,
    OnFirstRelease,
}
```

**Comment by jtroo** [comment]
Looks good to me

**Comment by jtroo** [comment]
The code has been updated to reflect our discussed changes.

**Comment by ald0o** [comment]
What is the status of this PR?
I pulled the code and I am currently using it.

It works quite fine in most cases, but I believe the logic isn't quite right in some corner cases. Indeed, one does not want the one-shot to be released until an actual related effect is triggered but, currently, if I press a one shot mod, then a hold-tap key, then the modifier is released before the hold-tap key has a chance to resolve to either the tap or the hold action.

In particular, in the case of a one-shot mod, there is no point in cancelling it before a key (or mouse) event is sent to host.
Currently, if I press a one shot mod, then a hold-tap key, then the modifier is released before the hold-tap key has a chance to resolve to either tap or hold action.

For a one-shot layer toggle, the logic is quite correct though, since every key can be affected by a layer change, even if it is not supposed to resolve as a keyboard event.

Maybe more fine-grained policies are required in the `OneShot` struct? (we need both a field for the press versus release policy and a field for the "all event" versus "side-effect events only" policy).

In this case, `Custom` actions also need a field to say whether they have side effects, since it cannot be guessed.

**Comment by TeXitoi** [comment]
Sorry I don't have so much time on keyberon recently. But as this MR is closed, it's not in my priority list.

**Comment by ald0o** [comment]
I assumed something like this!

So I will try to improve on it on my own and submit a new MR whenever I am done (unless @jtroo already addressed this in kanata-keyberon?).

**Comment by jtroo** [comment]
@ald0o 

The code in kanata-keyberon does have some fixes and works reasonably well, but is suited mostly for the use cases of kanata. The code has diverged in many ways, again for use cases suiting kanata, making it more challenging to make PRs and contribute back - hence why I closed this PR.

**Pull request closed without merging**